hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestListBlobProducer.java

```diff
@@ -59,65 +66,75 @@ public void setup() throws Exception {
-    configuration.set(FS_AZURE_MAX_CONSUMER_LAG, "10");
+    configuration.set(FS_AZURE_PRODUCER_QUEUE_MAX_SIZE, "10");
     AzureBlobFileSystem fs = (AzureBlobFileSystem) FileSystem.newInstance(
         configuration);
     AbfsClient client = fs.getAbfsClient();
     AbfsClient spiedClient = Mockito.spy(client);
     fs.getAbfsStore().setClient(spiedClient);
     fs.setWorkingDirectory(new Path("/"));
     fs.mkdirs(new Path("/src"));
+    ExecutorService executor = Executors.newFixedThreadPool(5);
+    List<Future> futureList = new ArrayList<>();
     for (int i = 0; i < 20; i++) {
-      fs.create(new Path("/src/file" + i));
+      int iter = i;
+      futureList.add(executor.submit(() -> {
+        return fs.create(new Path("/src/file" + iter));
+      }));
+    }
+    for(Future future : futureList) {
+      future.get();
     }
-    AtomicBoolean produced = new AtomicBoolean(true);
 
     AtomicInteger producedBlobs = new AtomicInteger(0);
     AtomicInteger listBlobInvoked = new AtomicInteger(0);
 
+    final ITestListBlobProducer testObj = this;
+    final ListBlobQueue queue = new ListBlobQueue(
+        fs.getAbfsStore().getAbfsConfiguration().getProducerQueueMaxSize(),
+        1);
+
     Mockito.doAnswer(answer -> {
-          listBlobInvoked.incrementAndGet();
-          AbfsRestOperation op = client.getListBlobs(answer.getArgument(0),
-              answer.getArgument(1), 1, answer.getArgument(3));
-          producedBlobs.incrementAndGet();
-          produced.set(true);
-          return op;
+      synchronized (testObj) {
+        listBlobInvoked.incrementAndGet();
+        AbfsRestOperation op = client.getListBlobs(answer.getArgument(0),
+            answer.getArgument(1), 1, answer.getArgument(3));
+        producedBlobs.incrementAndGet();
+        if(producedBlobs.get() > 10) {
+          Assert.assertTrue(queue.availableSize() > 0);
+        }
+        return op;
+      }
         })
         .when(spiedClient)
         .getListBlobs(Mockito.nullable(String.class),
             Mockito.nullable(String.class), Mockito.nullable(Integer.class),
             Mockito.nullable(TracingContext.class));
 
-    ListBlobQueue queue = new ListBlobQueue(null);
+
     ListBlobProducer producer = new ListBlobProducer("src/", spiedClient, queue,
         null, Mockito.mock(
         TracingContext.class));
     ListBlobConsumer consumer = new ListBlobConsumer(queue);
     while (producedBlobs.get() < 10) ;
 
-    int producedBlobCount = producedBlobs.get();
-
     int oldInvocation = listBlobInvoked.get();
-    Thread.sleep(10_000L);
     Assert.assertTrue(listBlobInvoked.get() == oldInvocation);
 
     while (!consumer.isCompleted()) {
-      produced.set(false);
-      consumer.consume();
-      while (!produced.get() && !queue.getIsCompleted()) ;
-      if (!queue.getIsCompleted()) {
-        Assert.assertEquals(producedBlobs.get() - 1, producedBlobCount);
+      synchronized (testObj) {
```

Comment:
Same here

Reply:
comment as above and part of same test. Reasoning as above.